File: app/src/main/java/com/tunjid/androidbootstrap/adapters/TileAdapter.java
package com.tunjid.androidbootstrap.adapters;

import android.view.ViewGroup;

import com.tunjid.androidbootstrap.R;
import com.tunjid.androidbootstrap.model.Tile;
import com.tunjid.androidbootstrap.recyclerview.InteractiveAdapter;
import com.tunjid.androidbootstrap.viewholders.TileViewHolder;

import java.util.List;

import androidx.annotation.NonNull;

/**
 * Adapter for BLE devices found while sacnning
 */
public class TileAdapter extends InteractiveAdapter<TileViewHolder, TileAdapter.AdapterListener> {

    private final List<Tile> tiles;

    public TileAdapter(List<Tile> tiles, AdapterListener scanAdapterListener) {
        super(scanAdapterListener);
        setHasStableIds(true);
        this.tiles = tiles;
    }

    @NonNull
    @Override
    public TileViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return new TileViewHolder(getItemView(R.layout.viewholder_tile, viewGroup), adapterListener);
    }

    @Override
    public void onBindViewHolder(@NonNull TileViewHolder viewHolder, final int position) {
        viewHolder.bind(tiles.get(position));
    }

    @Override
    public long getItemId(int position) {
        return tiles.get(position).getNumber();
    }

    @Override
    public int getItemCount() {
        return tiles.size();
    }

    public interface AdapterListener extends InteractiveAdapter.AdapterListener {
        void onTileClicked(Tile tile);
    }
}

File: app/src/main/java/com/tunjid/androidbootstrap/adapters/TileAdapter.kt
package com.tunjid.androidbootstrap.adapters

import android.view.ViewGroup

import com.tunjid.androidbootstrap.R
import com.tunjid.androidbootstrap.model.Tile
import com.tunjid.androidbootstrap.recyclerview.InteractiveAdapter
import com.tunjid.androidbootstrap.view.util.inflate
import com.tunjid.androidbootstrap.viewholders.TileViewHolder

class TileAdapter(private val tiles: List<Tile>, listener: (tile: Tile) -> Unit)
    : InteractiveAdapter<TileViewHolder, (tile: Tile) -> Unit>(listener) {

    init {
        setHasStableIds(true)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): TileViewHolder =
            TileViewHolder(viewGroup.inflate(R.layout.viewholder_tile), adapterDelegate)

    override fun onBindViewHolder(viewHolder: TileViewHolder, position: Int) =
            viewHolder.bind(tiles[position])

    override fun onViewRecycled(holder: TileViewHolder) {
        super.onViewRecycled(holder)
        holder.unBind()
    }

    override fun getItemId(position: Int): Long = tiles[position].number.toLong()

    override fun getItemCount(): Int = tiles.size

}
